pandas/_libs/groupby.pyx

```diff
@@ -703,7 +739,7 @@ def group_mean(floating[:, ::1] out,
             for j in range(K):
                 val = values[i, j]
                 # not nan
-                if val == val:
+                if val == val and not (is_datetimelike and val == NPY_NAT):
```

Comment:
Similar comment here - would rather be explicit about the cast that's occurring to make this work (eg hopefully this is int equality, and not float equality since comparing floats is problematic). 

On a brief investigation, one potential flaw here is that with `nan_val = NPY_NAT`, it holds true not only that `val == NPY_NAT`, but also that `val == NPY_NAT + 1`

Reply:
We have an implicit guarantee that datetimelike arrays will be returned as [`int64` in calls to the function](https://github.com/pandas-dev/pandas/blob/177f1abcf5cb50885484290bcfa8f236afc8767a/pandas/core/groupby/ops.py#L492.).

We can prevent any problems via asserting that `is_datetimelike=True` arrays come as `int64` at the top of the function. Going to try to do that and also add a test for `NPY_NAT + 1` just to be sure. Sounds good?

Reply:
I don't think that is possible because the type `floating` does not include `int64`, which is why I am assumed that the data is being cast to float before being passed to the algo.

A test for `NPY_NAT +1` would be great, if that's all good, then happy here (also fine with this as is, `NPY_NAT` handling is probably more an important than the edge case of `NPY_NAT + `

Reply:
If there's an issue with `NPY_NAT + 1`, could also just see if that gets fixed by changing the fused type to also include `int64_t` so that we can accept ints

Reply:
Yes, I was wrong. There is a `float64` cast down the line which occurs because [get_cython_func_and_vals is called a bit further down](https://github.com/pandas-dev/pandas/blob/177f1abcf5cb50885484290bcfa8f236afc8767a/pandas/core/groupby/ops.py#L511) which runs [`ensure_float64`](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/ops.py#L205) and actually does a pointless copy to float64 operation of the whole input array https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/algos_common_helper.pxi.in#L70.

We could change that path through the code to not do the cast for is_datetimelike but that would necessitate changing the input values data type to a union type which includes int64. Re-using `add_t` seems the best way to go since it would bring us one step closer to re-using group_add as suggested.

There is some risk through the union data type though, If I look at the algorithm of `group_add`, for one, it has extra-handling for the `object` dtype). However, that's nothing that can't be handled with an assert statement and an additional unit test.

What do you think?

Reply:
I think `groupby_t` would actually be a good candidate (`add_t` doesn't appear to even include `int64`). Hopefully just including those types won't necessitate much change (and can use `_treat_as_na` for more unified missing value handling since fused types will be the same (for example looking in `group_min_max`

Reply:
> What do you think?

1) If we find a workable solution for GroupBy.mean, we should also try to use it for nanops.nanmean (used by DTI, Series, and DataFrame mean methods)
2) if we stop casting to float64, we'll need to be extra-careful about silent overflows.
3) I think it would be reasonable to fix the NaT-specific problem separately since the PR already presents a good solution for that